source/extensions/common/wasm/context.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <atomic>
+#include <map>
+#include <memory>
+
+#include "include/proxy-wasm/context.h"
+
+#include "envoy/access_log/access_log.h"
+#include "envoy/buffer/buffer.h"
+#include "envoy/http/filter.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "common/common/assert.h"
+#include "common/common/logger.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+class Wasm;
+
+#include "proxy_wasm_common.h"
+
+using proxy_wasm::BufferInterface;
+using proxy_wasm::MetricType;
+using proxy_wasm::Pairs;
+using proxy_wasm::PairsWithStringValues;
+using proxy_wasm::PluginBase;
+using proxy_wasm::WasmBase;
+using proxy_wasm::WasmBufferType;
+using proxy_wasm::WasmHandle;
+using proxy_wasm::WasmResult;
+using proxy_wasm::WasmVm;
+
+using PluginBaseSharedPtr = std::shared_ptr<PluginBase>;
+
+class Buffer : public proxy_wasm::BufferInterface {
+public:
+  Buffer() {}
+
+  // Add additional wrappped types here.
+  Buffer* set(absl::string_view data) {
+    data_ = data;
+    return this;
+  }
+
+  size_t size() const override { return data_.size(); }
+  bool copyTo(WasmBase* wasm, size_t start, size_t length, uint64_t ptr_ptr,
+              uint64_t size_ptr) const override;
+
+private:
+  absl::string_view data_;
+};
+
+// Plugin contains the information for a filter/service.
+struct Plugin : public PluginBase {
+  Plugin(absl::string_view name, absl::string_view root_id, absl::string_view vm_id,
+         absl::string_view plugin_configuration,
+         const envoy::config::core::v3::TrafficDirection& direction,
+         const LocalInfo::LocalInfo& local_info,
+         const envoy::config::core::v3::Metadata* listener_metadata)
```

Comment:
Same regarding listener_metadata?

Reply:
Added comments and a TODO as discussed offline.